connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java

```diff
@@ -161,6 +164,7 @@ public void start() {
 
         // Create the topic admin client and initialize the topic ...
         admin = topicAdminSupplier.get();   // may be null
+        useAdminForListOffsets = admin != null;
```

Comment:
It's not obvious to me why we have to wait to call `topicAdminSupplier.get();` in `start`. 

I think we can call it in the constructor and not save the field `topicAdminSupplier` but keep only `admin` as a member field. Unfortunately it can't be declared `final` because it may be reset (set to `null`) in the exception handling below but still, using fewer member fields seems reasonable and does not reduce the usability in this class IMO.

Reply:
I don't think the wait time matters whether we call `topicAdminSupplier.get()` in the constructor or here, because the constructor and the `start()` method are called synchronously from the Connect & MM2 startup process.

However, I do think there is a lot of value in following the existing pattern of not actually gathering, creating or using resources until the `start()` method. And we still need all of the fields, including the `useAdminForListOffset` boolean field that is set in the start method but also may be changed later on in the `readEndOffsets(...)` method if the admin is talking to a broker that doesn't support the end offsets method (e.g., 0.10.x). In fact, the latter is the primary error condition this PR is trying to fix.

So, I think it's better to keep the code as is. But please let me know if you still disagree.

Reply:
Per the next comment, I've removed the `useAdminForListOffset` field to simplify things. However, I still think that it's more appropriate and follows existing convention for this class to call `topicAdminSupplier.get()` in the `start()` method.

Reply:
My main observation here is that keeping initialization in the constructor allows us to declare read only fields as `final`. 
It also offers better exception safety, in the sense that when an object is constructed (the call to its constructor succeeds) then we know that more or less we have a functional instance of that class. 

I'm not ignoring the cases where the initialization of variables has to happen at a later time during the call of `start`. But hopefully these cases are minimal, or else `start` ends up being the actual constructor of objects like this one here. Which can be redundant and in some cases more risky. 

As I mentioned, since `admin` is not immutable here (can be set to `null` later to support connection with older brokers) I'm fine keeping this call in `start`. But I'd still recommend not pushing initializations outside the constructor if they can happen at the time of the object construction (as is the case with `admin`'s initial initialization here).